doc/rtd/explanation/failure_states.rst

```diff
@@ -0,0 +1,126 @@
+.. _failure_states:
+
+Failure states
+==============
+
+Critical failure
+----------------
+If cloud-init is unable to complete, error messages will now
+additionally be visible in output of `cloud-init status --long`
+within the 'error' key nested under the module-level keys: 'init-local',
+'init', 'modules-config', 'modules-final'.
+
+Recoverable failure
+-------------------
+In the case that cloud-init is able to complete yet something goes awry,
+the service will now return with exit code 2, and error messages will be
+visible in the output of `cloud-init status --long` under the top
+level 'recoverable_errors' key as well as within the 'error' key nested
+under the module-level keys: 'init-local', 'init', 'modules-config',
+'modules-final'.
+
+Implementation
+==============
+
+Cloud-init error codes
+----------------------
+0 - success
```

Comment:
> I get the impression this might also need some introductory text - without context I'm a bit confused about whether these are universal error codes that are applied to all failures, or if they're more specific than that and only apply in certain cases.

+1, will do

> We may also want to indent these as a quote block to avoid them all being shown on the same line. My preference would be to have them in a bulleted list, with the code (and meaning) on the bulleted line in monospaced font, then the corresponding text under that bullet point on the next line (to make it clear where it belongs).

I think this comment applies to a prior version of this change. Could you please let me know if the updated version looks better?